beacon_node/network/src/sync/block_sidecar_coupling.rs

```diff
@@ -235,30 +321,61 @@ impl<E: EthSpec> RangeBlockComponentsRequest<E> {
+                    });
                 };
 
                 let mut custody_columns = vec![];
+                let mut naughty_peers = vec![];
                 for index in expects_custody_columns {
-                    let Some(data_column) = data_columns_by_index.remove(index) else {
-                        return Err(format!("No column for block {block_root:?} index {index}"));
-                    };
-                    // Safe to convert to `CustodyDataColumn`: we have asserted that the index of
-                    // this column is in the set of `expects_custody_columns` and with the expected
-                    // block root, so for the expected epoch of this batch.
-                    custody_columns.push(CustodyDataColumn::from_asserted_custody(data_column));
+                    if let Some(data_column) = data_columns_by_index.remove(index) {
+                        // Safe to convert to `CustodyDataColumn`: we have asserted that the index of
+                        // this column is in the set of `expects_custody_columns` and with the expected
+                        // block root, so for the expected epoch of this batch.
+                        custody_columns.push(CustodyDataColumn::from_asserted_custody(data_column));
+                    } else {
+                        // Penalize the peer for claiming to have the columns but not returning
+                        // them
+                        let Some(responsible_peer) = column_to_peer.get(index) else {
+                            return Err(CouplingError {
+                                msg: format!("Internal error, no request made for column {}", index),
+                                column_and_peer: None,
+                            });
+                        };
+                        naughty_peers.push((*index, *responsible_peer));
+                    }
+                }
+                if !naughty_peers.is_empty() {
+                    return Err(CouplingError {
+                        msg: format!("Peers did not return column for block_root {block_root:?} {naughty_peers:?}"),
+                        column_and_peer: Some((naughty_peers, PeerAction::LowToleranceError)),
+                    });
                 }
 
                 // Assert that there are no columns left
                 if !data_columns_by_index.is_empty() {
                     let remaining_indices = data_columns_by_index.keys().collect::<Vec<_>>();
-                    return Err(format!(
-                        "Not all columns consumed for block {block_root:?}: {remaining_indices:?}"
-                    ));
+                    // log the error but don't return an error, we can still progress with extra columns.
+                    tracing::error!(
+                        ?block_root,
+                        ?remaining_indices,
+                        "Not all columns consumed for block"
+                    );
```

Comment:
So this error can be triggered by bad peer behaviour, and there's nothing the user can do. It's not an internal error, I won't log anything or at most a `debug!` log

Reply:
Good call, I'll downgrade.